File: providers/aws/src/main/java/cloudspec/aws/ec2/resource/nested/EC2UserIdGroupPair.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.resource.nested;

import cloudspec.annotation.AssociationDefinition;
import cloudspec.annotation.PropertyDefinition;
import cloudspec.aws.ec2.resource.EC2SecurityGroupResource;
import cloudspec.aws.ec2.resource.EC2VpcPeeringConnectionResource;
import cloudspec.aws.ec2.resource.EC2VpcResource;
import software.amazon.awssdk.services.ec2.model.UserIdGroupPair;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class EC2UserIdGroupPair {
    @AssociationDefinition(
            name = "group",
            description = "The security group",
            targetClass = EC2SecurityGroupResource.class
    )
    private final String groupId;

    @PropertyDefinition(
            name = "group_name",
            description = "The name of the security group"
    )
    private final String groupName;

    @PropertyDefinition(
            name = "peering_status",
            description = "The status of a VPC peering connection, if applicable"
    )
    private final String peeringStatus;

    @PropertyDefinition(
            name = "userId",
            description = "The ID of an AWS account"
    )
    private final String userId;

    @AssociationDefinition(
            name = "vpc",
            description = "The VPC for the referenced security group, if applicable",
            targetClass = EC2VpcResource.class
    )
    private final String vpcId;

    @AssociationDefinition(
            name = "vpc_peering_connection",
            description = "The VPC peering connection, if applicable",
            targetClass = EC2VpcPeeringConnectionResource.class
    )
    private final String vpcPeeringConnectionId;

    public EC2UserIdGroupPair(String groupId, String groupName, String peeringStatus, String userId, String vpcId,
                              String vpcPeeringConnectionId) {
        this.groupId = groupId;
        this.groupName = groupName;
        this.peeringStatus = peeringStatus;
        this.userId = userId;
        this.vpcId = vpcId;
        this.vpcPeeringConnectionId = vpcPeeringConnectionId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (o == null || (getClass() != o.getClass() && !(o instanceof UserIdGroupPair))) {
            return false;
        }

        if (o instanceof UserIdGroupPair) {
            return sdkEquals((UserIdGroupPair) o);
        }

        EC2UserIdGroupPair that = (EC2UserIdGroupPair) o;
        return Objects.equals(groupId, that.groupId) &&
                Objects.equals(groupName, that.groupName) &&
                Objects.equals(peeringStatus, that.peeringStatus) &&
                Objects.equals(userId, that.userId) &&
                Objects.equals(vpcId, that.vpcId) &&
                Objects.equals(vpcPeeringConnectionId, that.vpcPeeringConnectionId);
    }

    private boolean sdkEquals(UserIdGroupPair that) {
        return Objects.equals(groupId, that.groupId()) &&
                Objects.equals(groupName, that.groupName()) &&
                Objects.equals(peeringStatus, that.peeringStatus()) &&
                Objects.equals(userId, that.userId()) &&
                Objects.equals(vpcId, that.vpcId()) &&
                Objects.equals(vpcPeeringConnectionId, that.vpcPeeringConnectionId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(groupId, groupName, peeringStatus, userId, vpcId, vpcPeeringConnectionId);
    }

    public static List<EC2UserIdGroupPair> fromSdk(List<UserIdGroupPair> userIdGroupPairs) {
        return Optional.ofNullable(userIdGroupPairs)
                       .orElse(Collections.emptyList())
                       .stream()
                       .map(EC2UserIdGroupPair::fromSdk)
                       .collect(Collectors.toList());
    }

    public static EC2UserIdGroupPair fromSdk(UserIdGroupPair userIdGroupPair) {
        return Optional.ofNullable(userIdGroupPair)
                       .map(v -> new EC2UserIdGroupPair(
                                       v.groupId(),
                                       v.groupName(),
                                       v.peeringStatus(),
                                       v.userId(),
                                       v.vpcId(),
                                       v.vpcPeeringConnectionId()
                               )
                       )
                       .orElse(null);
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/nested/EC2UserIdGroupPair.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2.nested

import cloudspec.annotation.AssociationDefinition
import cloudspec.annotation.PropertyDefinition
import cloudspec.aws.ec2.EC2SecurityGroup
import cloudspec.aws.ec2.EC2Vpc
import cloudspec.aws.ec2.EC2VpcPeeringConnection

data class EC2UserIdGroupPair(
        @property:AssociationDefinition(
                name = ASSOC_GROUP,
                description = ASSOC_GROUP_D,
                targetClass = EC2SecurityGroup::class
        )
        val groupId: String?,

        @property:PropertyDefinition(
                name = ASSOC_GROUP_NAME,
                description = ASSOC_GROUP_NAME_D
        )
        val groupName: String?,

        @property:PropertyDefinition(
                name = PROP_PEERING_STATUS,
                description = PROP_PEERING_STATUS_D
        )
        val peeringStatus: String?,

        @property:PropertyDefinition(
                name = PROP_USER_ID,
                description = PROP_USER_ID_D
        )
        val userId: String?,

        @property:AssociationDefinition(
                name = ASSOC_VPC,
                description = ASSOC_VPC_D,
                targetClass = EC2Vpc::class
        )
        val vpcId: String?,

        @property:AssociationDefinition(
                name = ASSOC_VPC_PEERING_CONNECTION,
                description = ASSOC_VPC_PEERING_CONNECTION_D,
                targetClass = EC2VpcPeeringConnection::class
        )
        val vpcPeeringConnectionId: String?
) {
    companion object {
        const val ASSOC_GROUP = "group"
        const val ASSOC_GROUP_D = "The security group"
        const val ASSOC_GROUP_NAME = "group_name"
        const val ASSOC_GROUP_NAME_D = "The name of the security group"
        const val PROP_PEERING_STATUS = "peering_status"
        const val PROP_PEERING_STATUS_D = "The status of a VPC peering connection, if applicable"
        const val PROP_USER_ID = "userId"
        const val PROP_USER_ID_D = "The ID of an AWS account"
        const val ASSOC_VPC = "vpc"
        const val ASSOC_VPC_D = "The VPC for the referenced security group, if applicable"
        const val ASSOC_VPC_PEERING_CONNECTION = "vpc_peering_connection"
        const val ASSOC_VPC_PEERING_CONNECTION_D = "The VPC peering connection, if applicable"
    }
}
